### PR TITLE
[CHM 287] migrate Km notes

### PR DESCRIPTION
**Pull request by liyuanmontreal**
migrate km-notes component.

component contains a textarea box and a history notes list.

### CODE REVIEW

File: src/components/inputs/notes.vue [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            :disabled="$attrs.disabled"
+            :rows="$attrs.rows" 
+            :cols="$attrs.cols" 
+            @change="save" >
+        </textarea>       
+   
+        <table class="table table-striped" v-if="notes?.length">
+            <tr v-for="(item, index) in notes" :key="index">
+                <td>{{item}}  </td>   
+                <td style="width:10px;"><span class="btn" @click="remove(index)"><i class="fa fa-trash-o icon"></i></span></td>
```
Comment (blaisefonseca) [inline src/components/inputs/notes.vue]: use bootstrap https://getbootstrap.com/docs/4.0/utilities

File: src/components/inputs/notes.vue [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            return  (JSON.parse(model.value)).filter(e=>e!=lastNote);
+        return [];
+   }); 
+   
+   const remove=(index)=>{  
+        for (var i = 0; i < notes.value.length; i++) {        
+            if(notes.value[i] == notes.value[index]){             
+                notes.value.splice(i, 1);        
+            }
+        }
```
Comment (blaisefonseca) [inline src/components/inputs/notes.vue]: `notes.value.splice(index, 1);` this should be enough

File: src/components/inputs/notes.vue [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    const save=()=>{   
+        const timestamp = new Intl.DateTimeFormat('en-US', {
+            year: 'numeric',
+            month: 'short',
+            day: '2-digit',
+            hour: 'numeric',
+            minute: 'numeric',
+            second: 'numeric',
+            hour12: true
+        }).format(new Date());
```
Comment (blaisefonseca) [inline src/components/inputs/notes.vue]: remove this and just use the filter datetime

File: src/components/inputs/notes.vue [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    tr:nth-child(odd)  span {
+        background-color:#E3E3E3
+    }
+    tr:nth-child(even)   span {
+        background-color:#EFEFEF
+    }
+    tr:nth-child(even)    {
+        background-color:#EFEFEF
+    }
```
Comment (blaisefonseca) [inline src/components/inputs/notes.vue]: do we need this custom styling?

File: index.vue [hunk 5]
```diff
@@ -77,7 +74,7 @@
   //<--------------- Testing Area ----------------------> 
   //Cop-count-down testing entries
   const endDate = new Date('2024-04-20');
-  const copUrl = 'https://www.unep.org/un-biodiversity-conference-cop-15';
+  const copUrl = 'https://www.unep.org/un-biodiversity-conferencnodee-cop-15';
```
Comment (sbilodeau) [inline index.vue]: useless changes

File: index.vue [hunk 6]
```diff
@@ -53,7 +49,8 @@
   import ScbdFooter from './src/components/cbd-nav/footer.vue'
   import CopCountDown from './src/components/countdown/cop-count-down.vue';
   import dateSelector from  './src/components/inputs/dateSelector.vue'
-  import { ref, provide } from 'vue';
+  import kmnotes from  './src/components/inputs/notes.vue'
```
Comment (sbilodeau) [inline index.vue]: shoudl use camel casing
```suggestion
  import kmNotes from  './src/components/inputs/notes.vue'
```

File: src/components/inputs/notes.vue [hunk 7]
```diff
@@ -0,0 +1,108 @@
+<template>
+    <div>
+        <textarea            
+            v-model="newNote"           
+            :id="$attrs.id"
+            :class="$attrs.class"
+            :required="$attrs.required" 
+            :placeholder="$attrs.placeholder"
+            :disabled="$attrs.disabled"
+            :rows="$attrs.rows"
```
Comment (sbilodeau) [inline src/components/inputs/notes.vue]: I would reccomande to make `rows` a props to control the default vale (to  `3` or `4`)

File: src/components/inputs/notes.vue [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        </table>   
+    </div> 
+</template>
+
+<script setup>
+   import { ref, computed } from 'vue'
+   import {useUser} from '../../services/composables/useAuth'
+  
+   let lastNote = '';
+   const newNote = ref();
```
Comment (sbilodeau) [inline src/components/inputs/notes.vue]: use empty string as defualt value
```suggestion
   const newNote = ref("");
```

File: src/components/inputs/notes.vue [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            month: 'short',
+            day: '2-digit',
+            hour: 'numeric',
+            minute: 'numeric',
+            second: 'numeric',
+            hour12: true
+        }).format(new Date());
+        const lNotes = notes.value;
+        if(lNotes[lNotes.length-1] == lastNote)
+             lNotes.splice(lNotes.length-1, 1);
```
Comment (sbilodeau) [inline src/components/inputs/notes.vue]: No need complexify the coding with lengh etc. 
simply us at() and pop().  at() with negative value start from the end of the array

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop

```suggestion
        if(lNotes.at(-1) == lastNote)
             lNotes.pop();
```

File: src/components/inputs/notes.vue [hunk 10]
```diff
@@ -0,0 +1,108 @@
+<template>
+    <div>
+        <textarea            
+            v-model="newNote"           
+            :id="$attrs.id"
+            :class="$attrs.class"
```
Comment (sbilodeau) [inline src/components/inputs/notes.vue]: default class to bootstrp  `form-control`

File: src/components/inputs/notes.vue (hunk 4 above)
Comment (sbilodeau) [reply src/components/inputs/notes.vue]: I think we shoudl simply use bootstrap `table-stripped` and `fomr-control` for the textarea

File: src/components/inputs/notes.vue [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            hour: 'numeric',
+            minute: 'numeric',
+            second: 'numeric',
+            hour12: true
+        }).format(new Date());
+        const lNotes = notes.value;
+        if(lNotes[lNotes.length-1] == lastNote)
+             lNotes.splice(lNotes.length-1, 1);
+            
+        var newLNote = newNote.value? "[ " +  user.value.name  + " | " + timestamp + " ] - " + newNote.value.trim():"";
```
Comment (sbilodeau) [inline src/components/inputs/notes.vue]: use string litreal tmeplate using backtick `
```suggestion
        const newLNote = newNote.value ? `[ ${user.value.name} | ${timestamp} ] - ${newNote.value.trim()}` : '';  
```

**Review by sbilodeau — CHANGES_REQUESTED** [review]
as general rules:

- use camelCase
- preferably use string templating wuth backtick ` instead of oncatenation using ""+""
- use boostrap classes for style and spacing (https://getbootstrap.com/docs/5.3/utilities/spacing/)

File: src/components/inputs/notes.vue (hunk 1 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: done

File: src/components/inputs/notes.vue (hunk 2 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: done

File: src/components/inputs/notes.vue (hunk 3 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: done

File: src/components/inputs/notes.vue (hunk 4 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: delete custom css; use table , table-striped for table, use form-control for textarea

remain issue: use table-striped doesn't show zebra-stripping effect

File: index.vue (hunk 5 above)
Comment (liyuanmontreal) [reply index.vue]: misoperation , fixed

File: index.vue (hunk 6 above)
Comment (liyuanmontreal) [reply index.vue]: done

File: src/components/inputs/notes.vue (hunk 10 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: done

File: src/components/inputs/notes.vue (hunk 11 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: done

File: src/components/inputs/notes.vue (hunk 9 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: done

File: src/components/inputs/notes.vue (hunk 8 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: done

File: src/components/inputs/notes.vue (hunk 7 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: done

File: index.vue [hunk 12]
```diff
@@ -53,7 +48,8 @@
   import ScbdFooter from './src/components/cbd-nav/footer.vue'
   import CopCountDown from './src/components/countdown/cop-count-down.vue';
   import dateSelector from  './src/components/inputs/dateSelector.vue'
-  import { ref, provide } from 'vue';
+  import kmNotes from  './src/components/inputs/notes.vue'
```
Comment (sbilodeau) [inline index.vue]: the notes must be located in components/controls/ or components/controls/km. not in 'input' it's not a basic field type

**Review by sbilodeau — CHANGES_REQUESTED** [review]
Did not realzied but file badly located

File: src/components/inputs/notes.vue (hunk 4 above)
Comment (liyuanmontreal) [reply src/components/inputs/notes.vue]: fix table stripped issue

File: index.vue (hunk 12 above)
Comment (liyuanmontreal) [reply index.vue]: change to componentss/controls path

File: index.vue [hunk 13]
```diff
@@ -104,4 +104,4 @@
   filter: drop-shadow(0 0 2em #42b883aa);
 }
 </style>
-  
\ No newline at end of file
+  ./src/components/controls/notes.vue
```
Comment (blaisefonseca) [inline index.vue]: why do we need this path here?

**Review by blaisefonseca — APPROVED** [review]

**Pull request merged**